src/hotspot/share/prims/jvmtiRawMonitor.cpp

```diff
@@ -410,38 +375,45 @@ int JvmtiRawMonitor::raw_wait(jlong millis, Thread* self) {
   OrderAccess::fence();
 
   intptr_t save = _recursions;
-  _recursions = 0;
-  _waiters++;
-  ret = simple_wait(self, millis);
-  _recursions = save;
-  _waiters--;
-
-  guarantee(self == _owner, "invariant");
 
   if (self->is_Java_thread()) {
     JavaThread* jt = self->as_Java_thread();
     guarantee(jt->thread_state() == _thread_in_native, "invariant");
-    for (;;) {
-      if (!SafepointMechanism::should_process(jt)) {
-        // Not suspended so we're done here:
-        break;
+
+    _recursions = 0;
+    _waiters++;
+    ret = simple_wait(self, millis);
+    _waiters--;
```

Comment:
Good catch! Even if the block wasn't moved we still no longer own the monitor after simple_wait.

This is going to be tricky to fix in a clear/clean way as the waiters count has to be maintained properly.

Reply:
Actually the field is unused.
The _entry_list is the list of waiters, which is the only thing we really care about.
If there was a reader of _waiter reading it without lock can only be used as some kind of hint.
If you have lock you can just check _entry_list.

I duplicated it because otherwise I would need two branchse on if (self->is_Java_thread()) which creates more complicated code.